### PR TITLE
Support minimum_should_match for terms_set query

### PR DESCRIPTION
**Pull request by Kiriakos1998**
Support minimum_should_match for terms_set query. 

Closes(#94095)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 1]
```diff
@@ -82,6 +85,7 @@ protected void doWriteTo(StreamOutput out) throws IOException {
         out.writeGenericValue(values);
         out.writeOptionalString(minimumShouldMatchField);
         out.writeOptionalWriteable(minimumShouldMatchScript);
+        out.writeOptionalString(minimumShouldMatch);
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:88]: This needs to handle bwc serialization. 

There are multiple examples around the code base. 

Something like:
```
if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_500_019)) {
```
^ Where you create the next transport version type in that class with a UUID

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 2]
```diff
@@ -74,6 +76,7 @@ public TermsSetQueryBuilder(StreamInput in) throws IOException {
         this.values = (List<?>) in.readGenericValue();
         this.minimumShouldMatchField = in.readOptionalString();
         this.minimumShouldMatchScript = in.readOptionalWriteable(Script::new);
+        this.minimumShouldMatch = in.readOptionalString();
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:79]: This needs to handle bwc serialization. 

There are multiple examples around the code base. 

Something like:
```
if (in.getTransportVersion().onOrAfter(TransportVersion.V_8_500_019)) {
..setvalue
}
```
^ Where you create the next transport version type in that class with a UUID

**Review by benwtrent — CHANGES_REQUESTED** [review]
Sorry for taking so long on a review. Please add BWC handling and then we can see where CI fails :)

**Comment by Kiriakos1998** [comment]
Hello @benwtrent, just pushed the changes handling bwc serialization. Let me know if anything else is needed.

**Comment by benwtrent** [comment]
Jenkins retest this please

**Comment by Kiriakos1998** [comment]
> Jenkins retest this please

It failed in changelog validation. I will fix it and push the changes.

**Comment by benwtrent** [comment]
Jenkins retest this please

**Review by benwtrent — COMMENTED** [review]
OK, reviewed some more. I like the tests and the logic all seems good to me. 

Once you update the error message && the transport version conflict is fixed (it looks like `main` has a new version and you need to bump yours and handle the conflict) I think this is ready to merge.

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 3]
```diff
@@ -110,24 +120,37 @@ public Script getMinimumShouldMatchScript() {
...
         return this;
     }
 
+    public String getMinimumShouldMatch() {
+        return minimumShouldMatch;
+    }
+
+    public TermsSetQueryBuilder setMinimumShouldMatch(String minimumShouldMatch) {
+        if (minimumShouldMatchField != null || minimumShouldMatchScript != null) {
+            throw new IllegalArgumentException("A field has already been specified. Cannot specify both a field and script");
```
[the remark this answers is not in the document]
Comment (Kiriakos1998) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:136]: So we need something like that: "A field or a script has already been specified. Cannot specify both a custom field and, a field or script" and accordingly for the other error messages.

[the remark this answers is not in the document]
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:136]: The new minimum should match option isn't a field. It's just a value. So maybe say "value, field, or script"?

[the remark this answers is not in the document]
Comment (Kiriakos1998) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java]: Just pushed the changes. Updated error messages and added new TransportVersion

**Comment by benwtrent** [comment]
Jenkins retest this please

**Comment by Kiriakos1998** [comment]
> Jenkins retest this please

I think the version was updated that's why the tests are failing.  After rebasing with the main should be ok.

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 4]
```diff
@@ -44,15 +44,19 @@ public final class TermsSetQueryBuilder extends AbstractQueryBuilder<TermsSetQue
 
     public static final String NAME = "terms_set";
 
+    public static final TransportVersion MINIMUM_SHOULD_MATCH_ADDED_VERSION = TransportVersion.V_8_500_041;
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:47]:
```suggestion
    public static final TransportVersion MINIMUM_SHOULD_MATCH_ADDED_VERSION = TransportVersion.V_8_500_047;
```

**Comment by benwtrent** [comment]
Jenkins retest this please

File: server/src/main/java/org/elasticsearch/TransportVersion.java [hunk 5]
```diff
@@ -169,9 +169,10 @@ private static TransportVersion registerTransportVersion(int id, String uniqueId
     public static final TransportVersion V_8_500_044 = registerTransportVersion(8_500_044, "96b83320-2317-4e9d-b735-356f18c1d76a");
     public static final TransportVersion V_8_500_045 = registerTransportVersion(8_500_045, "24a596dd-c843-4c0a-90b3-759697d74026");
     public static final TransportVersion V_8_500_046 = registerTransportVersion(8_500_046, "61666d4c-a4f0-40db-8a3d-4806718247c5");
+    public static final TransportVersion V_8_500_047 = registerTransportVersion(8_500_041, "4b1682fe-c37e-4184-80f6-7d57fcba9b3d");
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/TransportVersion.java:172]:
```suggestion
    public static final TransportVersion V_8_500_047 = registerTransportVersion(8_500_047, "4b1682fe-c37e-4184-80f6-7d57fcba9b3d");
```

**Comment by benwtrent** [comment]
Jenkins retest this please

**Review by benwtrent — APPROVED** [review]

**Comment by benwtrent** [comment]
PR merged!
🎉 🎉 🎉 @Kiriakos1998 🎉 🎉 🎉 

Thanks for your contribution!!!

**Comment by Kiriakos1998** [comment]
> PR merged! 🎉 🎉 🎉 @Kiriakos1998 🎉 🎉 🎉
> 
> Thanks for your contribution!!!

You are welcome.

**Pull request merged**